### DISCUSSION TITLE
Allow additional authentication for /superset/sql_json endpoint

### DISCUSSION THREAD

**Discussion by sewardgw** [apache/superset#14182] [Q&A / Help]
The `/superset/sql_json` endpoint provides a great way to externally invoke arbitrary SQL through Superset which manages manages the credentials, driver, libraries, and DB connections, etc. This is useful for a number of use-cases where external systems want to delegate the DB interaction; one common example is integration with Amundsen. However, in order to use this endpoint the client seems to have 2 options for authentication:

1. Provide a valid csrf token
2. Set `WTF_CSRF_ENABLED=False` in a custom `superset_config.py` (or whitelist the endpoint)

For these use-cases the authentication should be programatic which makes option 1 not possible and, while option 2 works, this configuration should ideally not need to be turned off.

**Describe the solution you'd like**
Be able to authenticate and use the API without csrf tokens.

Possibly have an `/api/v1` verstion of the `sql_json` endpoint that leverages bearer token auth.

**Describe alternatives you've considered**
Open to ideas!

**Additional context**
Happy to know of other ways to accomplish this if it already exists.

**Comment by nytai** [comment]
I think the ideal fix here is to migrate this api to `/api/v1/. We already have an implementation for most major apis except for SQL Lab. Porting the sql_json endpoint over to a SQLRestApi implementation should be pretty straightforward since it's already a json api. Though, it could benefit from some refactor. @betodealmeida or @dpgaspar are either of you aware of any plans to add a sql endpoint to the api? 

@sewardgw disabling csrf protection on the endpoint _should_ be ok for now, as json apis are [not subject to csrf attacks](https://github.com/dxa4481/CORS) in most cases.

**Comment by sewardgw** [comment]
Gotcha, good to know!

Do you know what the proper config should be in `superset_config.py` to white listing this one endpoint? I had this in my file but it didn't seem to pick up properly:

```python
WTF_CSRF_EXEMPT_LIST = ['superset.views.core.sql_json']
```

**Reply by HenrryVargas** [reply]
Hi @sewardgw 

I've same issue. Did you resolved?

I am trying to connect Amundsen to Superset and the Superset log show:
POST //superset/sql_json/ HTTP/1.1 401 51 python-requests/2.25.1

I have  superset_config in this way:


```
import os
from cachelib.redis import RedisCache

def env(key, default=None):
    return os.getenv(key, default)

MAPBOX_API_KEY = env('MAPBOX_API_KEY', '')
CACHE_CONFIG = {
      'CACHE_TYPE': 'redis',
      'CACHE_DEFAULT_TIMEOUT': 300,
      'CACHE_KEY_PREFIX': 'superset_',
      'CACHE_REDIS_HOST': env('REDIS_HOST'),
      'CACHE_REDIS_PORT': env('REDIS_PORT'),
      'CACHE_REDIS_PASSWORD': env('REDIS_PASSWORD'),
      'CACHE_REDIS_DB': env('REDIS_DB', 1),
}
DATA_CACHE_CONFIG = CACHE_CONFIG

SQLALCHEMY_DATABASE_URI = f"postgresql+psycopg2://{env('DB_USER')}:{env('DB_PASS')}@{env('DB_HOST')}:{env('DB_PORT')}/{env('DB_NAME')}"
SQLALCHEMY_TRACK_MODIFICATIONS = True
SECRET_KEY = env('SECRET_KEY', 'thisISaSECRET_1234')

# Flask-WTF flag for CSRF
WTF_CSRF_ENABLED = False
# Add endpoints that need to be exempt from CSRF protection
WTF_CSRF_EXEMPT_LIST = ['superset.views.core.sql_json']
# A CSRF token that expires in 1 year
WTF_CSRF_TIME_LIMIT = 60 * 60 * 24 * 365
class CeleryConfig(object):
  BROKER_URL = f"redis://{env('REDIS_HOST')}:{env('REDIS_PORT')}/0"
  CELERY_IMPORTS = ('superset.sql_lab', )
  CELERY_RESULT_BACKEND = f"redis://{env('REDIS_HOST')}:{env('REDIS_PORT')}/0"
  CELERY_ANNOTATIONS = {'tasks.add': {'rate_limit': '10/s'}}

CELERY_CONFIG = CeleryConfig
RESULTS_BACKEND = RedisCache(
      host=env('REDIS_HOST'),
      port=env('REDIS_PORT'),
      key_prefix='superset_results'
)

```

**Comment by nytai** [comment]
that should definitely work... strange. Here's where it's applied, if you're interested in debugging 
https://github.com/apache/superset/blob/df7e2b6a8e5ce01cbc79d832e86fc4b8511f4385/superset/app.py#L686
